src/crypto/merkle.rs

```diff
@@ -93,36 +97,58 @@ const EMPTY_ROOTS: [Hash; MAX_TREE_HEIGHT] = [
     hex!("33ebfb34d3aa119cb665d564acdd77b318dc86aa6a97744edb3bc03e97d776ca"),
 ];
 
+/// Marker trait for the leaf nodes of a Merkle tree.
+pub trait MerkleLeaf: AsRef<[u8]> {}
+
+/// Marker trait for the root of a Merkle tree.
+pub trait MerkleRoot: From<Hash> + Into<Hash> + Clone {}
+
+/// Marker trait for the proof of a Merkle tree.
+pub trait MerkleProof: AsRef<[Hash]> + From<Vec<Hash>> {}
+
+impl MerkleLeaf for Vec<u8> {}
+impl MerkleRoot for Hash {}
+impl MerkleProof for Vec<Hash> {}
+
+pub type PlainMerkleTree = MerkleTree<Vec<u8>, Hash, Vec<Hash>>;
+pub type SliceMerkleTree = MerkleTree<Vec<u8>, Hash, Vec<Hash>>;
+pub type DoubleMerkleTree = MerkleTree<Hash, Hash, Vec<Hash>>;
+
 /// Implementation of a Merkle tree.
-pub struct MerkleTree {
+pub struct MerkleTree<L: AsRef<[u8]>, R: MerkleRoot, P: MerkleProof> {
     /// All hashes in the tree, leaf hashes and inner nodes.
     nodes: Vec<Hash>,
     /// For each level, has the offset in `nodes` and the number of hashes on that level.
     levels: Vec<(u32, u32)>,
+    /// Marker for the type of the tree.
+    _type: PhantomData<(L, R, P)>,
 }
 
-impl MerkleTree {
+impl<L: AsRef<[u8]>, R: MerkleRoot, P: MerkleProof> MerkleTree<L, R, P> {
     /// Creates a new Merkle tree from the given data for each leaf.
     ///
     /// This will always create a perfect binary tree (filling with empty leaves as necessary).
     /// If you want to create a tree with more than half of the leaves empty,
     /// you have to explicitly pass in empty leaves as part of `data`.
-    pub fn new(data: &[impl AsRef<[u8]>]) -> Self {
-        Self::new_from_iter(data.iter().map(|leaf| leaf.as_ref()))
+    pub fn new(leaves: &[L]) -> Self {
```

Comment:
nit: seems to only be used in tests.  I wonder if it will be a lot of work to update the tests to use `new_from_iter()` instead.